libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -1,79 +1,141 @@
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun restart(route: NavigationRoute, onRoutesRefreshed: (List<NavigationRoute>) -> Unit) {
-        stop()
-        if (route.routeOptions.enableRefresh() == true) {
-            routerRefreshTimer.startTimer {
-                refreshRoute(route, onRoutesRefreshed)
+    private suspend fun tryRefreshingUntilRouteChanges(
+        initialRoute: NavigationRoute
+    ): NavigationRoute {
+        while (true) {
+            val refreshed = refreshRoute(initialRoute)
+            if (refreshed != initialRoute) {
+                return refreshed
             }
         }
     }
 
-    /**
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun stop() {
-        routerRefreshTimer.stopJobs()
-        currentRequestId?.let {
-            directionsSession.cancelRouteRefreshRequest(it)
-            currentRequestId = null
+    private suspend fun refreshRoute(
+        route: NavigationRoute
+    ): NavigationRoute = coroutineScope {
+        val routeLegs = route.directionsRoute.legs()
+        require(routeLegs != null) { "Can't refresh route without legs" }
+
+        var timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+        for (attempts in 0..2) {
+            timeUntilNextAttempt.await()
+            timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+            val refreshedRoute = withTimeoutOrNull(routeRefreshOptions.intervalMillis) {
+                refreshRouteOrNull(route)
+            }
+            if (refreshedRoute != null) {
+                timeUntilNextAttempt.cancel()
+                return@coroutineScope refreshedRoute
+            }
         }
+        timeUntilNextAttempt.cancel()
+        removeTrafficAnnotations(route, routeLegs)
     }
 
-    private fun refreshRoute(
+    private fun removeTrafficAnnotations(
         route: NavigationRoute,
-        onRoutesRefreshed: (List<NavigationRoute>) -> Unit
-    ) {
+        routeLegs: List<RouteLeg>
+    ): NavigationRoute {
+        val currentLegIndex = currentLegIndexProvider()
+        return route.refreshRoute(
+            0,
+            routeLegs.mapIndexed { index, leg ->
+                if (index >= currentLegIndex) {
+                    leg.annotation()
+                        ?.toBuilder()
+                        ?.congestion(leg.annotation()?.congestion()?.map { "unknown" })
+                        ?.congestionNumeric(leg.annotation()?.congestionNumeric()?.map { null })
+                        ?.build()
+                } else {
+                    leg.annotation()
+                }
+            },
+            routeLegs.mapIndexed { index, it ->
+                if (index >= currentLegIndex) {
+                    it.incidents()?.filter {
+                        val parsed = parseISO8601DateToLocalTimeOrNull(it.endTime())
+                            ?: return@filter false
```

Comment:
```suggestion
                            ?: return@filter true
```
Do we want to clear incidents that have no `endTime`? I imagine we want to leave them be.

Reply:
I thought it's better to remove them if we aren't sure. Do you think we should keep incidents if we couldn't parse the end time?

Reply:
Yes, I think we should remove something only if we're certain it's optimal to do so. Otherwise, let's leave the data in there.

Reply:
done